src/LinearSolverTypes.f90

```diff
@@ -1431,17 +1427,53 @@ SUBROUTINE getResidual_LinearSolverType_Iterative(solver,resid)
     IF(resid%n == solver%b%n) THEN
 #ifdef HAVE_MKL
       !not yet implemented
-      SELECTTYPE(b => solver%b); TYPE IS(RealVectorType)
-        resid%b=-b%b
+      SELECTTYPE(b => solver%b)
+      TYPE IS(RealVectorType)
+        SELECTTYPE(resid)
+        TYPE IS(RealVectorType)
+          resid%b=-b%b
+          CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid,ALPHA=1.0_SRK,BETA=1.0_SRK)
+        CLASS DEFAULT
+          CALL eLinearSolverType%raiseError('Incorrect call to '// &
+              modName//'::getResidual'//' - Type mismatch between getResidual argument and solver datatypes')
+        ENDSELECT
+#ifdef HAVE_MPI
+      TYPE IS(NativeDistributedVectorType)
+        SELECTTYPE(resid)
+        TYPE IS(NativeDistributedVectorType)
+          resid%b=-b%b
+          CALL BLAS_matvec(THISMATRIX=solver%A,X=solver%X,Y=resid,ALPHA=1.0_SRK,BETA=1.0_SRK)
+        CLASS DEFAULT
+          CALL eLinearSolverType%raiseError('Incorrect call to '// &
+              modName//'::getResidual'//' - Type mismatch between getResidual argument and solver datatypes')
+        ENDSELECT
+#endif
       ENDSELECT
```

Comment:
Should have a `CLASS DEFAULT` statement that throws an error

Reply:
This routine's been updated. The getresidual method's been changed around to use the more general BLAS_matvec routines.
It also gets rid of a lot of headaches when integrating with CMFD; basically there's an interface now which controls whether you get iteration count and residual norm or just the residual vector. It's also implemented for the direct solvers so you don't have to select type iterative all the time.